Issue #1039: Use browser-engine-system and feature-session components to replace WebView abstraction layer..

pocmo opened the pull request:
@mcomella @Baron-Severin I want to share the current state of the migration to use `browser-engine`. This version is functional and you can use it to browse the web. Like with the `browser-session` integration I used extension functions to work around functionality that is not available in the upstream components.

Missing pieces without a workaround in this version (list may be incomplete):
* **Tracking protection** toggle not functional (TP is already in components 0.20, with 0.21 [Friday] we get the option to configure this globally for an engine or session and that will make this toggle simpler to implement.
* We can't **intercept URLs** to inject HTML currently (used for firefox:home, firefox:about and error pages). ~~This functionality is scheduled to land in 0.22 (next week). However I may move this into the current sprint so that we can have this in 0.21 on Friday.~~ We moved this into the 0.21 sprint and the functionality should be available in 0.21.
  * Without loading "firefox:home" and adding it to the tab's history this currently breaks some assumptions when using the back button. This should fix itself once "firefox:home" gets loaded normally.
* **Fullscreen API** (works with YouTube because the TV website doesn't use the web API and just shows the content at full size). Scheduled to land in 0.21 on Friday.

Added follow-up items to: #1063 

---

As a side effect:
 * Replaces AmazonWebView with WebView (Issue #379)
 * Removes the product flavor dimensions that are not needed anymore.

pocmo committed:
Issue #1039: Use browser-engine-system and feature-session components to replace WebView abstraction layer.

As a side effect:
 * Replaces AmazonWebView with WebView (Issue #379)
 * Removes the product flavor dimensions that are not needed anymore.

mcomella commented on app/src/main/java/org/mozilla/focus/Components.kt:
```diff
@@ -4,38 +4,19 @@
-    override fun name(): String {
-        throw NotImplementedError()
-    }
+    val sessionUseCases = SessionUseCases(sessionManager)
```
nit: by lazy? Otherwise engine and sessionManager will make themselves right now.

pocmo replied:
Oh, yeah, good catch!

mcomella reviewed (commented):
With a quick skim, this looks reasonable to me. We can spend more time reviewing when we land for real (next week?). Note: I'll be PTO next week so I'll be unable to look over this until the following Monday.

![image](https://user-images.githubusercontent.com/759372/44882636-0463e800-ac69-11e8-8290-b21b51116b62.png)

You win Fire TV!

mcomella commented:
> Note: I'll be PTO next week so I'll be unable to look over this until the following Monday.

Unless we can land something tomorrow. :)

pocmo commented:
> @pocmo How much is changing for the final PR?

For this iteration not much.

Components 0.21 should get released today and then with that I should be able to look at the three missing pieces mentioned above. The patches for tracking protection and URL interception landed in a-c. Fullscreen API is at risk for 0.21 because the PR is not up yet. We may need to ship that with 0.22 or can ship a 0.21.1 if this gets in early next week.

> Provided the changes aren't too big, I can look over the most recent WIP tomorrow, Severin can look at it next week and land it, and I can look at it to make sure we didn't miss anything when I come back. 

Depending on what you and @Baron-Severin think I'd be open to land this as-is (with review comments addressed) and open separate PRs for the next pieces. The idea is to unblock others by having to wait on a big code drop and to be able to have others test and iterate on the new code. The drawback is that we temporarily have a not 100% feature complete app though (-> at least the three points mentioned above). Apart from that the app seems to be "fully" functional in my tests.

pocmo commented:
Updated the PR. Made all checks pass (ktlint, detekt, checkstyle, lint), SessionUseCases initializes lazily now and cleaned up some of the code.

mcomella reviewed (approved):
A few things missing here or there and a few nits but otherwise lgtm. I find the `SessionUseCases` abstraction strange: see below.

Some other things that have changed – please file a bug or fix:
- Back/forward state takes a while to update

Not sure it was related to your changes but I got into a state where a white page was loaded and the progress bar was shown without a URL (i.e. empty str or null) after going  backwards and forwards a few times.

> I'd be open to land this as-is (with review comments addressed) and open separate PRs for the next pieces.

Me too: please send an email to the team letting them know 1) this is landing, 2) the features that need follow-ups, and 3) that they'll need to change their build variant. Also, don't forget to file issues to track this work! :)

@Baron-Severin FYI that this may land.

pocmo commented on app/build.gradle:
```diff
@@ -269,11 +237,11 @@ def addSecretToBuildConfig(variant, fieldName, fileBaseName) {
     // accidentally ship a release using the debug key or vice versa.
     def fileSuffix
-    if (variantName.contains("Debug")) {
+    if (variantName.contains("debug")) {
```
Yeah, now that one dimension is gone they are just "debug" etc..

pocmo commented on app/src/main/java/org/mozilla/focus/ext/EngineView.kt:
```diff
@@ -0,0 +1,182 @@
+        isHorizontalScrollBarEnabled = true
+    }
+
+    webView.settings.apply {
```
Makes sense. I'll add that.

pocmo commented on app/src/amazonWebview/java/org/mozilla/focus/webview/FirefoxAmazonWebView.kt:
```diff
@@ -1,225 +0,0 @@
-    }
-
-    @Suppress("UNUSED_PARAMETER") // To avoid allocation, we use a function reference but it's on a platform interface.
-    private fun onFocusChange(view: View, hasFocus: Boolean): Unit = if (!hasFocus) {
```
Added this to list #1063.

pocmo commented on app/src/main/java/org/mozilla/focus/ext/EngineView.kt:
```diff
@@ -0,0 +1,182 @@
+    }
+}
+
+fun EngineView.scrollByClamped(vx: Int, vy: Int) {
```
Added this to list in #1063.

pocmo commented on app/src/main/java/org/mozilla/focus/browser/BrowserFragment.kt:
```diff
@@ -119,9 +117,9 @@ class BrowserFragment : IWebViewLifecycleFragment(), Session.Observer {
-            NavigationEvent.BACK -> if (webView?.canGoBack() ?: false) webView?.goBack()
-            NavigationEvent.FORWARD -> if (webView?.canGoForward() ?: false) webView?.goForward()
-            NavigationEvent.TURBO, NavigationEvent.RELOAD -> webView?.reload()
+            NavigationEvent.BACK -> if (session.canGoBack) requireComponents.sessionUseCases.goBack.invoke()
```
I saw you filed https://github.com/mozilla-mobile/android-components/issues/697. Let's discuss there.

pocmo commented:
From the migration I feel like the overlay and the things touching the engine are pretty well separated.  We are definitely introducing risk with the migration but it doesn't feel like it gets worse by having both things (engine migration + overlay migration) happen at the same time.

> whose components haven't been used in a production app before

Yeah, that's a chicken egg problem. FFFTV is the first candidate because the feature set is the smallest and the first milestone we have reached. :)

> there are regressions in back/forward state that I couldn't reproduce consistently 

I noticed that there is a regression because we can't intercept "firefox:home" and inject custom content. With that the page is not in the history stack and breaks some of the logic. The functionality for that landed in 0.21 - I added an item about testing back/forward after that to #1063.

> Perhaps we should land but not hesitate to back it out if it causes problems? What do you think @pocmo ?

Yeah, that sounds good to me. It will get more complicated to back out later on after we added more commits on top of it though.

I'll talk to QA once this lands. :)

pocmo commented:
@Baron-Severin From my side this is ready to land. I addressed most of @mcomella's review comments, rebased the branch and added follow-up items to #1063. I started working on the follow-up items now (using 0.21 of the components) but I'll create separate PRs for those changes once this PR lands.

pocmo commented on app/src/main/java/org/mozilla/focus/ext/Fragment.kt:
```diff
@@ -0,0 +1,11 @@
+import android.support.v4.app.Fragment
+import org.mozilla.focus.Components
+
+val Fragment.requireComponents: Components
```
It's not shorter but I think it shows the intent better? If I see `context!!` I'm never sure whether someone just wanted to make the code compile or if they actually *assume* that this will not be null here. Naming-wise I followed `requireContext` and `requireActivity` from Fragment. And I noticed I should use `requireContext` here... :)

pocmo commented on app/src/main/java/org/mozilla/focus/utils/IntentUtils.java:
```diff
@@ -88,10 +88,10 @@ public static boolean handleExternalUri(final Context context, final IWebView we
         final String fallbackUrl = intent.getStringExtra(EXTRA_BROWSER_FALLBACK_URL);
         if (fallbackUrl != null) {
-            webView.loadUrl(fallbackUrl);
+            //TODO: webView.loadUrl(fallbackUrl);
```
Oh, I missed that part. I'll update the code.

pocmo commented on app/src/main/java/org/mozilla/focus/iwebview/EngineViewLifecycleFragment.kt:
```diff
@@ -10,64 +10,50 @@ import android.support.annotation.UiThread
 
     /** Get the initial URL to load after the view has been created. */
     abstract val initialUrl: String
     abstract val session: Session
```
Ah, nice, thanks! I'll refactor that. :)

pocmo commented on app/src/main/java/org/mozilla/focus/ext/EngineView.kt:
```diff
@@ -0,0 +1,182 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+@file:Suppress("UNUSED_PARAMETER")
```
Ah, this shouldn't be necessary anymore. I needed that during the migration when I had not implemented all extension methods yet. :)

pocmo commented on app/src/main/java/org/mozilla/focus/browser/InfoFragment.java:
```diff
@@ -1,118 +0,0 @@
-
-        final String url = getInitialUrl();
-        if (!(url.startsWith("http://") || url.startsWith("https://"))) {
-            // Hide webview until content has loaded, if we're loading built in about/rights/etc
```
Added to #1063.

pocmo commented on app/src/main/java/org/mozilla/focus/browser/InfoFragment.java:
```diff
@@ -1,118 +0,0 @@
-        return new IWebView.Callback() {
-            @Override
-            public void onPageStarted(final String url) {
-                progressView.announceForAccessibility(getString(R.string.accessibility_announcement_loading));
```
Added to #1063.

pocmo commented on app/src/main/java/org/mozilla/focus/browser/InfoActivity.java:
```diff
@@ -1,99 +0,0 @@
-            final View view = WebViewProvider.create(this, attrs, MainActivity.getFactory());
-
-            final IWebView webView = (IWebView) view;
-            webView.setBlockingEnabled(false);
```
Added to #1063.

severinrudie reviewed (approved):
Had a few questions, nothing major.  LGTM

pocmo commented on app/build.gradle:
```diff
@@ -13,7 +13,7 @@ android {
 
     defaultConfig {
-        applicationId "org.mozilla.tv"
+        applicationId "org.mozilla.tv.firefox"
```
Previously we had product flavors for the engine and those flavors added a suffix:

```Groovy
roductFlavors {
        amazonWebview {
            applicationIdSuffix ".firefox"
            // [..]
        }
        gecko {
            applicationIdSuffix ".gecko"
            // [..]
        }
    }
```

We have been shipping the `amazonWebview` flavor only. Now that the flavor is gone I moved the suffix to the `applicationId` setup directly.

pocmo commented on app/src/main/java/org/mozilla/focus/browser/InfoActivity.kt:
```diff
@@ -0,0 +1,83 @@
+        val actionBar = supportActionBar
+        if (actionBar != null) {
+            actionBar.setDisplayHomeAsUpEnabled(true)
+            actionBar.setDisplayShowHomeEnabled(true)
```
That's a good idea and avoids some repetition. I'll update the PR.
